google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java

```diff
@@ -699,6 +717,30 @@ public boolean isNull(int columnIndex) {
       return rowData.get(columnIndex) == null;
     }
 
+    @Override
+    protected <T extends AbstractMessage> T getProtoMessageInternal(int columnIndex, T message) {
+      Preconditions.checkNotNull(
+          message,
+          "Proto message may not be null. Use MyProtoClass.getDefaultInstance() as a parameter value.");
+      try {
+        return (T)
+            message
+                .toBuilder()
+                .mergeFrom(((LazyByteArray) rowData.get(columnIndex)).getByteArray().toByteArray())
+                .build();
+      } catch (InvalidProtocolBufferException e) {
+        throw SpannerExceptionFactory.asSpannerException(e);
+      }
```

Comment:
I think that this would be more efficient, as we would then not copy the byte array around as many times. Instead, it just creates a stream of bytes from the `base64` string that can be decoded directly.

```suggestion
      try {
        return (T)
            message
                .toBuilder()
                .mergeFrom(
                    Base64.getDecoder()
                        .wrap(
                            CharSource.wrap(((LazyByteArray) rowData.get(columnIndex)).base64String)
                                .asByteSource(StandardCharsets.UTF_8)
                                .openStream()))
                .build();
      } catch (IOException ioException) {
        throw SpannerExceptionFactory.asSpannerException(ioException);
      }
```

Reply:
Thanks for suggesting this optimization, @olavloite. I agree with you here that we are copying the byte array multiple times which could affect performance when proto messages are big.
I have made the changes in the same PR.